libs/data/bbhnet/data/transforms/whitening.py

```diff
@@ -88,17 +113,75 @@ def forward(self, X: torch.Tensor) -> torch.Tensor:
         X = X.transpose(2, 0)
         X = X - X.mean(axis=0)
         X = X.transpose(0, 2)
-        X *= self.window
+
+        self.pad = int(np.ceil(self.time_domain_filter.size(-1) / 2))
+
+        X[:, :, : self.pad] *= self.window[: self.pad]
+        X[:, :, -self.pad :] *= self.window[-self.pad :]
 
         # convolve the detrended data with the time-domain
         # filters constructed during initialization from
         # the background data, using groups to ensure that
         # the convolution is performed independently for
         # each interferometer channel
-        X = torch.nn.functional.conv1d(
-            X, self.time_domain_filter, groups=self.num_ifos, padding=self.pad
-        )
+
+        # see
+        # https://github.com/gwpy/gwpy/blob/main/gwpy/timeseries/timeseries.py
+        # for logic
+
+        nfft = min(8 * self.time_domain_filter.size(-1), self.kernel_size)
+
+        # number of samples of corrupted data
+        # due to settling in of whitening filter
+        crop_samples = int((self.fduration / 2) * self.sample_rate)
+
+        if nfft >= self.kernel_size / 2:
+            conv = torch.nn.functional.conv1d(
+                X,
+                self.time_domain_filter,
+                groups=self.num_ifos,
+                padding="same",
+            )
+
+            # crop the beginning and ending fduration / 2
+            conv = conv[:, :, crop_samples:-crop_samples]
+
+        # else use the overlap-save algorithm
+        else:
+            nstep = nfft - 2 * self.pad
+            conv = torch.zeros_like(X)
+            # handle first chunk separately
+            conv[:, :, : nfft - self.pad] = torch.nn.functional.conv1d(
+                X[:, :, :nfft],
+                self.time_domain_filter,
+                groups=self.num_ifos,
+                padding="same",
+            )[:, :, : nfft - self.pad]
+
+            # process chunks of length nstep
+            k = nfft - self.pad
+            while k < self.kernel_size - nfft + self.pad:

```

Comment:
Looping and doing multiple convolutions is going to be pretty inefficient on GPU, especially since each convolution will be comparatively small because it only operates on a subslice (basically you're going to waste time launching a bunch of kernels for ops that will be pretty small, while GPUs can parallelize individual operations really efficiently). Consider using the torch [unfold](https://pytorch.org/docs/stable/generated/torch.nn.Unfold.html) operation up front, which will basically create multiple windowed views of the timeseries at a given stride (which for reasons having to do with what this op is intended for, I believe you'll want to pass as the `dilation` argument, _not_ the `stride` argument).

Then you can do a single convolution along the last axis, slice off the pads so that the views are non-overlapping, then do a simple reshape. This solution might only work for timeseries whose length can be divided evenly into chunks like this, but it's probably worth enforcing this for efficiency, especially at inference time.

Reply:
Yeah that sounds great